Documentation/developers/debugging.rst

```diff
@@ -0,0 +1,99 @@
+    - ``critical``: any error that is forcing a shutdown of the service or application to prevent data loss (or further data loss). These errors should be reserved only for the most heinous errors and situations where there is guaranteed to have been data corruption or loss.
+
+Each Polycube service has a ``loglevel`` property that enables the printing of log messages filtered by the choosen level.
+Setting the ``loglevel`` to ``X`` means that all the messages of that loglevel and above will be printed.
+For instance, if ``loglevel=warning``, all messages marked as ``warning``, ``error`` and ``critical`` are printed.
+
+Setting the ``loglevel`` property of a service is documented in :ref:`polycubectl section <complex-conf>`.
+
+
+**Note 1**: the loglevel of ``polycubectl`` does not affect the outputs produced by the logs of individual services. In other words, ``polycubectl`` loglevel can be set to ``info``, while the loglevel of the ``bridge`` service can be set to ``trace``.
+
+**Note 2**: the log file can be easily filtered on either (1) a specific loglevel or (2) a specific service by using the standard Unix ``grep`` tool.
+
+    Usage examples:
+
+    ::
+
+        sudo polycubed | grep "debug"
+        sudo polycubed | grep -E "warning|error|critical"
+        sudo polycubed | grep "bridge"
+
+.. _logging-in-the-data-plane:
+
+Logging in the data plane
+***************************************
+In order to avoid the use of the ``bpf_trace_printk()`` primitive (which is not integrated with control plane logging and should not be used in Polycube), the Polycube framework provides the ``pcn_log()`` primitive.
+
+Syntax:
+
+::
+
+    pcn_log(ctx, level, msg, args...)
+
+- ``ctx``: packet being processed.
+- ``level``: type of logging level: ``LOG_TRACE``, ``LOG_DEBUG``, ``LOG_INFO``, ``LOG_WARN``, ``LOG_ERR``, ``LOG_CRITICAL``.
+- ``msg``: message to print
+- ``args``: arguments of the message
+
+Three special format specifiers are available:
+- ``%I``: print an IP address
+- ``%M``: print a MAC address
+- ``%P``: print a TCP/UDP port
+
+Please note the the custom specifiers expect the data to be in *network byte order* while standard specifiers expects it to be in *host byte order*.
+
+Current limitations of ``pcn_log()``:
+ - Cannot be used inside a macro
+ - Maximum 4 arguments are allowed
+
+Usage example:
+
+::
+
+  pcn_log(ctx, LOG_DEBUG, "Receiving packet from port %d", md->in_port);
+
+
+The ``pcn_pkt_log(ctx, level)`` primitive sends a packet to the control plane where it is printed in a tcpdump-like format.
+`ctx` and `level` are the same as in `pcn_log`.
+This feature is only designed for developers, so final version of services should not include this.
+
+`polycube-tools <https://github.com/mauriciovasquezbernal/polycube-tools>`_ must be installed and ``polycube`` cmake must be configured with `cmake -DHAVE_POLYCUBE_TOOLS=ON ..`
```

Comment:
@mauriciovasquezbernal Is this still valid? As far as I know, @michelsciortino was not able to make it working and he didn't understand exactly what this was about.

Reply:
Thanks, Mauricio.
@michelsciortino can you please add a new commit in which you:
1) update the documentation, removing all the references to polycube-tools;
2) update the code, as the "HAVE_POLYCUBE_TOOLS" macro is present in many makefiles and it should be removed.

Reply:
@frisso so this means logging with packet dumping in control plane not supported any more or there is alternatives?

Reply:
You can use the new "span" mode for a port, which enables to receive all packets on a new interface available in a dedicated namespace. There, you can use standard tools such as tcpdump/Wireshark to inspect the traffic.

Reply:
@michelsciortino You should re-organize the commits (using the git squash function) in order to make it easier to understand. Details by voice.

Reply:
@michelsciortino what's the status of this PR? Is the re-base done?